### ISSUE TITLE
Native file dialogs do not prevent input from propagating to main window

### ISSUE THREAD

**Issue by Yagich** [godotengine/godot#85047]
### Godot version

4.2rc1

### System information

Linux/Manjaro

### Issue description

`Windows`s with the `exclusive` property set to `true` are supposed to not propagate input to the main window. However, native dialogs, such as `FileDialog` (which is inherently `exclusive` by default) with `use_native_dialog` set to `true` still let input through.

When running the MRP, you can press 1 to popup a native dialog and press the button. If you press 2, it will popup a `Window` dialog and you will not be able to press the button.

### Steps to reproduce

1. Display a native dialog, either with `DisplayServer#file_dialog_show` or a `FileDialog` with `Use Native Dialog` enabled
2. Click or keyboard focus any Control in the tree while the dialog is showing

### Minimal reproduction project

[native-file-dialog-exclusive-bug.zip](https://github.com/godotengine/godot/files/13399881/native-file-dialog-exclusive-bug.zip)

**Comment by bruvzg** [comment]
This is expected. Native file dialogs are not part of the app itself and fully handled by OS, `exclusive` flag (or any other Window flag) can't have any effect on it.

In case of Linux it's part of [XDG desktop portal](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.FileChooser.html) and there's no option to make it exclusive (we are setting parent window XID, but it's only used to arrange dialog on top of the window).

**Comment by ondesic** [comment]
This would be great to implement if possible. The native dialog is superior to the filedialog node. We just need a way to make it exclusive

**Comment by nonchip** [comment]
A workaround is to have parent window ignore inputs (might be as simple as pausing the scenetree or as complex as actually doing logic on the kind of input we're blocking) while the dialog is visible, instead of relying on window management. this should also be relatively simple to implement in GDScript until a better solution is found.

**Comment by MewPurPur** [comment]
I can't find any way to register when you're canceling the native file dialog, and it's not a Godot node. So these workarounds don't work.

**Comment by nonchip** [comment]
@MewPurPur it's a Godot node: [`FileDialog`](https://docs.godotengine.org/en/latest/classes/class_filedialog.html). ~~from the docs i would expect it to still fake its cancel button (`get_cancel_button` on the parent class) signal when native, but didn't test.~~ (see bruvzg below)

there's also the [`DisplayServer`](https://docs.godotengine.org/en/latest/classes/class_displayserver.html#class-displayserver-method-file-dialog-show) method, which will call its callback when it closes with the first argument depending on whether the user clicked ok or cancel (not that that value actually matters for the logic, just the fact it closed -> fired its callback)

also tbh a dialog that you dont know whether it's open or not (or whether to do the action or cancel) sounds pretty much unusable, if that wasnt a thing, we'd have a way bigger problem / more prioritized issue.

(i also explained my idea to you in chat, specifically answering your question how to know when to unpause, before summarizing it here for posterity, bit confused why you'd now days later just claim it to be wrong)

i'm wondering now, if you didnt know about the callback nor the node, how did you get the result the user picked? is there another way i'm not seeing? I didnt notice any suspicious 1-element arrays being passed.

**Comment by bruvzg** [comment]
If `FileDialog` node is used to show native dialog, it should emit `canceled` signal (generic signal of the `AcceptDialog` node). If `DisplayServer` `file_dialog_with_options_show/file_dialog_show` method is used, it should call its callback function with first argument set to `false`.

**Comment by ondesic** [comment]
If we can open the native FileDialog and listen for the file selected signal, we must have some ability to listen to events. Can we get the ability to listen for visibility_changed, focus_exited, canceled and other signals so we can try and prevent users pressing outside the native file dialog?

**Comment by nonchip** [comment]
what do you mean by "get the ability to listen"? like mentioned already, both ways of opening a native file dialog do tell you (the node via an actual signal, the displayserver using the callback) when it closes.

also how would those other, unrelated (and probably unavailable, depending on WM) signals prevent input from reaching another window? simply have that other window ignore input while the dialog is open, if that's how it should behave.

**Comment by ondesic** [comment]
The other signals will help keep it visible. If you click out of the dialogbox, it closes without sending a signal. By seeing when it loses focus, changes visibility, we can make it visible again if it closes without the user pressing OK or cancel.

**Comment by nonchip** [comment]
> If you click out of the dialogbox, it closes without sending a signal.

that would be a bug then, and unrelated to input propagating to parent windows or loosing focus/changing visibility. [see above](https://github.com/godotengine/godot/issues/85047#issuecomment-3042322056) for what it should do depending how you opened it.

this issue is about making/faking exclusive native file dialogs.

**Comment by ondesic** [comment]
Yes, this is actually what I am saying. If the native dialog was exclusive, it would block signals sent to underneath nodes. that way it would not disappear when clicking outside the dialog while opened.
I can open another bug but calinou sent me here when I asked outside this post. I think the problem is related.

**Comment by nonchip** [comment]
> that way it would not disappear when clicking outside the dialog while opened.

you're talking about *wildly* different things there. this issue is about godot window behavior, you're talking about a feature/behavior in your platform's native dialog (and something that you cant just prevent from happening by reacting to some signal).

in fact, on my desktop environment, it *doesnt* disappear on its own when you click outside. and yes, the dialog closing without emitting any kind of response/signal/callback/whatever is in fact a bug unrelated to this issue (which, unlike yours, actually shouldnt be here to begin with imo because it's asking for a new feature instead of reporting a bug; or if anything it's reporting a documentation oversight).

**Comment by ondesic** [comment]
Ok, I can open another bug report. I just hope it doesn't get closed because someone says it is a duplicate of this thread.
